kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -282,6 +379,277 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
 
     val nonTransientBindings = bindings.filterNotNull()
     val options = JsonReader.Options.of(*nonTransientBindings.map { it.name }.toTypedArray())
-    return KotlinJsonAdapter(constructor, bindings, nonTransientBindings, options).nullSafe()
+    return KotlinJsonAdapter(constructorData, bindings, nonTransientBindings, options).nullSafe()
+  }
+
+  private infix fun KmType?.isEqualTo(other: KmType?): Boolean {
+    return when {
+      this === other -> true
+      this != null && other != null -> {
+        abbreviatedType == other.abbreviatedType
+            && arguments areEqualTo other.arguments
+            && classifier == other.classifier
+            && flags == other.flags
+            && flexibleTypeUpperBound isEqualTo other.flexibleTypeUpperBound
+            && outerType isEqualTo other.outerType
+      }
+      this == null && other == null -> true
+      else -> false
+    }
+  }
+
+  private infix fun Collection<KmTypeProjection>.areEqualTo(
+      other: Collection<KmTypeProjection>): Boolean {
+    // check collections aren't same
+    if (this !== other) {
+      // fast check of sizes
+      if (this.size != other.size) return false
+      val areNotEqual = this.asSequence()
+          .zip(other.asSequence())
+          // check this and other contains same elements at position
+          .map { (fromThis, fromOther) -> fromThis isEqualTo fromOther }
+          // searching for first negative answer
+          .contains(false)
+      if (areNotEqual) return false
+    }
+    // collections are same or they are contains same elements with same order
+    return true
+  }
+
+  private infix fun KmTypeProjection?.isEqualTo(other: KmTypeProjection?): Boolean {
+    return when {
+      this === other -> true
+      this != null && other != null -> {
+        variance == other.variance
+            && type isEqualTo other.type
+      }
+      this == null && other == null -> true
```

Comment:
I suppose this is already checked by `this === other -> true`?

Reply:
Hmm, good point! I added the this === other check later 🙃 d3e6e36